### PR TITLE
middleware support logics

### PR DESCRIPTION
**Pull request by kaibocai**
### Issue describing the changes in this PR

resolves #issue_for_this_pr

### Pull request checklist

* [X] My changes **do not** require documentation changes
  * [ ] Otherwise: Documentation issue linked to PR
* [X] My changes **should not** be added to the release notes for the next release
  * [ ] Otherwise: I've added my notes to `release_notes.md`
* [X] My changes **do not** need to be backported to a previous version
  * [ ] Otherwise: Backport tracked by issue/PR #issue_or_pr
* [X] I have added all required tests (Unit tests, E2E tests)

### Additional information

Additional PR information

### CODE REVIEW

File: src/main/java/com/microsoft/azure/functions/worker/binding/ExecutionContextDataSource.java [hunk 1]
```diff
@@ -17,6 +25,11 @@ public final class ExecutionContextDataSource extends DataSource<ExecutionContex
     private final BindingDataStore dataStore;
     private final MethodBindInfo methodBindInfo;
     private final Class<?> containingClass;
+    private final Map<String, Parameter> parameterMap = new HashMap<>();
+    private final Map<String, String> parameterPayloadMap = new HashMap<>();
+    private final Map<String, Object> middlewareInputMap = new HashMap<>();
+    private Object returnValue;
+    private Object middlewareOutput;
```
Comment (trask) [inline src/main/java/com/microsoft/azure/functions/worker/binding/ExecutionContextDataSource.java]: revisit

File: src/main/java/com/microsoft/azure/functions/worker/broker/JavaFunctionBroker.java [hunk 2]
```diff
@@ -64,6 +105,7 @@ private ExecutionContextDataSource buildExecutionContext(String id,  InvocationR
 		ExecutionContextDataSource executionContextDataSource = new ExecutionContextDataSource(request.getInvocationId(),
 				traceContext, retryContext, methodEntry.left, dataStore, functionDefinition.getCandidate(),
 				functionDefinition.getContainingClass());
+		executionContextDataSource.buildParameterPayloadMap(request.getInputDataList());
```
Comment (trask) [inline src/main/java/com/microsoft/azure/functions/worker/broker/JavaFunctionBroker.java]: pass into constructor to make state immutable

File: src/main/java/com/microsoft/azure/functions/worker/binding/ExecutionContextDataSource.java [hunk 3]
```diff
@@ -35,6 +48,7 @@ public ExecutionContextDataSource(String invocationId, TraceContext traceContext
         this.dataStore = dataStore;
         this.methodBindInfo = methodBindInfo;
         this.containingClass = containingClass;
+        addParameters(methodBindInfo, this.parameterMap);
```
Comment (trask) [inline src/main/java/com/microsoft/azure/functions/worker/binding/ExecutionContextDataSource.java]:
```suggestion
        this.parameterMap = createParameterMap(methodBindInfo);
```

File: src/main/java/com/microsoft/azure/functions/worker/binding/ExecutionContextDataSource.java [hunk 4]
```diff
@@ -64,4 +78,72 @@ public MethodBindInfo getMethodBindInfo() {
...
+            parameterMap.put(paramBindInfo.getName(), paramBindInfo.getParameter());
+        }
+    }
+
+    @Override
+    public Map<String, Parameter> getParameterMap() {
+        return this.parameterMap;
+    }
+
+    public void buildParameterPayloadMap(List<ParameterBinding> inputDataList){
```
Comment (trask) [inline src/main/java/com/microsoft/azure/functions/worker/binding/ExecutionContextDataSource.java]:
```suggestion
    private static Map<String, String> buildParameterPayloadMap(List<ParameterBinding> inputDataList){
```

File: src/main/java/com/microsoft/azure/functions/worker/binding/ExecutionContextDataSource.java [hunk 5]
```diff
@@ -64,4 +78,72 @@ public MethodBindInfo getMethodBindInfo() {
...
+        for (ParameterBinding parameterBinding : inputDataList) {
+            String serializedPayload = convertToString(parameterBinding.getData());
+            this.parameterPayloadMap.put(parameterBinding.getName(), serializedPayload);
+        }
+    }
+
+    // TODO: Refactor the code in V5 to make resolve arguments logs before middleware invocation
```
Comment (trask) [inline src/main/java/com/microsoft/azure/functions/worker/binding/ExecutionContextDataSource.java]:
```suggestion
    // TODO: Refactor the code in V5 to make resolve arguments logics before middleware invocation
```

**Review by trask — COMMENTED** [review]
from joint review with Kaibo and Sreyas

File: src/main/java/com/microsoft/azure/functions/worker/binding/ExecutionContextDataSource.java [hunk 6]
```diff
@@ -64,4 +83,91 @@ public MethodBindInfo getMethodBindInfo() {
...
+    public Optional<String> getParameterName(String name){
+        for (Map.Entry<String, Parameter> entry : this.parameterMap.entrySet()){
+            if (isOrchestrationTrigger(entry.getValue(), name)){
+                return Optional.of(entry.getKey());
+            }
+        }
+        return Optional.empty();
+    }
+
+    private static boolean isOrchestrationTrigger(Parameter parameter, String name){
```
Comment (trask) [inline src/main/java/com/microsoft/azure/functions/worker/binding/ExecutionContextDataSource.java]: rename method

File: src/main/java/com/microsoft/azure/functions/worker/binding/ExecutionContextDataSource.java [hunk 7]
```diff
@@ -17,15 +28,21 @@ public final class ExecutionContextDataSource extends DataSource<ExecutionContex
...
     private static final DataOperations<ExecutionContext, Object> EXECONTEXT_DATA_OPERATIONS = new DataOperations<>();
     static {
         EXECONTEXT_DATA_OPERATIONS.addGenericOperation(ExecutionContext.class, DataOperations::generalAssignment);
     }
 
     public ExecutionContextDataSource(String invocationId, TraceContext traceContext, RetryContext retryContext,
-                               String funcname, BindingDataStore dataStore, MethodBindInfo methodBindInfo,
-                               Class<?> containingClass){
+                                      String funcname, BindingDataStore dataStore, MethodBindInfo methodBindInfo,
+                                      Class<?> containingClass, List<ParameterBinding> inputDataList){
```
Comment (trask) [inline src/main/java/com/microsoft/azure/functions/worker/binding/ExecutionContextDataSource.java]:
```suggestion
                                      Class<?> containingClass, List<ParameterBinding> parameterBindings){
```

File: src/main/java/com/microsoft/azure/functions/worker/binding/ExecutionContextDataSource.java [hunk 8]
```diff
@@ -64,4 +83,91 @@ public MethodBindInfo getMethodBindInfo() {
...
+                return true;
+            }
+        }
+        return false;
+    }
+
+    private static Map<String, String> buildParameterPayloadMap(List<ParameterBinding> inputDataList){
+        Map<String, String> map = new HashMap<>();
+        for (ParameterBinding parameterBinding : inputDataList) {
+            String serializedPayload = convertToString(parameterBinding.getData());
```
Comment (trask) [inline src/main/java/com/microsoft/azure/functions/worker/binding/ExecutionContextDataSource.java]: only add String parameters to map

File: src/main/java/com/microsoft/azure/functions/worker/broker/ParameterResolver.java [hunk 9]
```diff
@@ -41,7 +41,10 @@ private static InvokeInfoBuilder resolve(ExecutionContextDataSource executionCon
                     argument = dataStore.getOrAddDataTarget(invokeInfo.getOutputsId(), paramName, paramType, false);
                 }
                 else if (paramName != null && !paramName.isEmpty()) {
-                    argument = dataStore.getDataByName(paramName, paramType);
+                    argument = buildMiddlewareInput(executionContextDataSource.getMiddlewareInputByName(paramName));
+                    if (!argument.isPresent()) {
+                        argument = dataStore.getDataByName(paramName, paramType);
+                    }
```
Comment (trask) [inline src/main/java/com/microsoft/azure/functions/worker/broker/ParameterResolver.java]:
```suggestion
                    argument = executionContextDataSource.getBindingData(paramName);
```

File: src/main/java/com/microsoft/azure/functions/worker/binding/ExecutionContextDataSource.java [hunk 10]
```diff
@@ -64,4 +83,91 @@ public MethodBindInfo getMethodBindInfo() {
...
+            case BYTES:  return data.getBytes().toString(StandardCharsets.UTF_8);
+            case JSON:   return data.getJson();
+            case HTTP:   return data.getHttp().toString();
+            case COLLECTION_STRING: data.getCollectionString().toString();
+            case COLLECTION_DOUBLE: data.getCollectionDouble().toString();
+            case COLLECTION_BYTES: data.getCollectionBytes().toString();
+            case COLLECTION_SINT64: data.getCollectionSint64().toString();
+            default: return null;
+        }
+    }
```
Comment (trask) [inline src/main/java/com/microsoft/azure/functions/worker/binding/ExecutionContextDataSource.java]:
```suggestion
```

File: src/main/java/com/microsoft/azure/functions/worker/binding/ExecutionContextDataSource.java [hunk 11]
```diff
@@ -64,4 +83,91 @@ public MethodBindInfo getMethodBindInfo() {
...
+    private static Map<String, Parameter> addParameters(MethodBindInfo methodBindInfo){
+        Map<String, Parameter> map = new HashMap<>();
+        for (ParamBindInfo paramBindInfo : methodBindInfo.getParams()) {
+            map.put(paramBindInfo.getName(), paramBindInfo.getParameter());
+        }
+        return map;
+    }
+
+    @Override
+    public Optional<String> getParameterName(String name){
```
Comment (trask) [inline src/main/java/com/microsoft/azure/functions/worker/binding/ExecutionContextDataSource.java]:
```suggestion
    public Optional<String> getParameterName(String annotationType){
```

File: src/main/java/com/microsoft/azure/functions/worker/binding/ExecutionContextDataSource.java [hunk 12]
```diff
@@ -64,4 +83,91 @@ public MethodBindInfo getMethodBindInfo() {
...
+            case COLLECTION_STRING: data.getCollectionString().toString();
+            case COLLECTION_DOUBLE: data.getCollectionDouble().toString();
+            case COLLECTION_BYTES: data.getCollectionBytes().toString();
+            case COLLECTION_SINT64: data.getCollectionSint64().toString();
+            default: return null;
+        }
+    }
+
+    @Override
+    public Object getParameterPayloadByName(String name) {
```
Comment (trask) [inline src/main/java/com/microsoft/azure/functions/worker/binding/ExecutionContextDataSource.java]:
```suggestion
    public Object getParameterValue(String name) {
```

File: src/main/java/com/microsoft/azure/functions/worker/binding/ExecutionContextDataSource.java [hunk 13]
```diff
@@ -64,4 +83,91 @@ public MethodBindInfo getMethodBindInfo() {
...
+    }
+
+    @Override
+    public void updateParameterPayloadByName(String key, Object value) {
+        this.middlewareInputMap.put(key, value);
+    }
+
+    public Object getMiddlewareInputByName(String name){
+        return this.middlewareInputMap.get(name);
+    }
```
Comment (trask) [inline src/main/java/com/microsoft/azure/functions/worker/binding/ExecutionContextDataSource.java]:
```suggestion
```

File: src/main/java/com/microsoft/azure/functions/worker/broker/ParameterResolver.java [hunk 14]
```diff
@@ -62,6 +65,11 @@ else if (paramName == null && !paramBindingNameAnnotation.isEmpty()) {
         }
     }
 
+    private static Optional<BindingData> buildMiddlewareInput(Object input) {
+        if (input == null) return Optional.empty();
+        return Optional.of(new BindingData(input));
+    }
```
Comment (trask) [inline src/main/java/com/microsoft/azure/functions/worker/broker/ParameterResolver.java]: (move)

```suggestion
```

File: src/main/java/com/microsoft/azure/functions/worker/binding/ExecutionContextDataSource.java [hunk 15]
```diff
@@ -17,15 +28,21 @@ public final class ExecutionContextDataSource extends DataSource<ExecutionContex
     private final BindingDataStore dataStore;
     private final MethodBindInfo methodBindInfo;
     private final Class<?> containingClass;
+    private final Map<String, Parameter> parameterMap;
+    private final Map<String, String> parameterPayloadMap;
+    private final Map<String, Object> middlewareInputMap = new HashMap<>();
+    private Object returnValue;
+    private Object middlewareOutput;
```
Comment (trask) [inline src/main/java/com/microsoft/azure/functions/worker/binding/ExecutionContextDataSource.java]: do we need this field? (combine with returnValue)

File: src/main/java/com/microsoft/azure/functions/worker/binding/ExecutionContextDataSource.java [hunk 16]
```diff
@@ -17,15 +28,21 @@ public final class ExecutionContextDataSource extends DataSource<ExecutionContex
     private final BindingDataStore dataStore;
     private final MethodBindInfo methodBindInfo;
     private final Class<?> containingClass;
+    private final Map<String, Parameter> parameterMap;
+    private final Map<String, String> parameterPayloadMap;
+    private final Map<String, Object> middlewareInputMap = new HashMap<>();
```
Comment (trask) [inline src/main/java/com/microsoft/azure/functions/worker/binding/ExecutionContextDataSource.java]: combine into single map? `parameterValues`

File: src/main/java/com/microsoft/azure/functions/worker/binding/ExecutionContextDataSource.java [hunk 17]
```diff
@@ -17,15 +28,21 @@ public final class ExecutionContextDataSource extends DataSource<ExecutionContex
     private final BindingDataStore dataStore;
     private final MethodBindInfo methodBindInfo;
     private final Class<?> containingClass;
+    private final Map<String, Parameter> parameterMap;
```
Comment (trask) [inline src/main/java/com/microsoft/azure/functions/worker/binding/ExecutionContextDataSource.java]:
```suggestion
    // key is ...
    private final Map<String, Parameter> parameterDefinitions;
```

File: src/main/java/com/microsoft/azure/functions/worker/binding/ExecutionContextDataSource.java [hunk 18]
```diff
@@ -64,4 +83,91 @@ public MethodBindInfo getMethodBindInfo() {
...
+    public void updateParameterPayloadByName(String key, Object value) {
+        this.middlewareInputMap.put(key, value);
+    }
+
+    public Object getMiddlewareInputByName(String name){
+        return this.middlewareInputMap.get(name);
+    }
+
+    public void setReturnValue(Object retValue) {
+        this.returnValue = retValue;
```
Comment (trask) [inline src/main/java/com/microsoft/azure/functions/worker/binding/ExecutionContextDataSource.java]: maybe

```suggestion
        this.returnValue = retValue;
        this.dataStore.setDataTargetValue(BindingDataStore.RETURN_NAME, this.middlewareOutput);
```

File: src/main/java/com/microsoft/azure/functions/worker/broker/EnhancedJavaMethodExecutorImpl.java [hunk 19]
```diff
@@ -28,6 +21,7 @@ public void execute(ExecutionContextDataSource executionContextDataSource) throw
                     .orElseThrow(() -> new NoSuchMethodException("Cannot locate the method signature with the given input"))
                     .invoke(() -> executionContextDataSource.getContainingClass().newInstance());
             executionContextDataSource.getDataStore().setDataTargetValue(BindingDataStore.RETURN_NAME, retValue);
+            executionContextDataSource.setReturnValue(retValue);
```
Comment (trask) [inline src/main/java/com/microsoft/azure/functions/worker/broker/EnhancedJavaMethodExecutorImpl.java]: combine these two lines?

File: src/main/java/com/microsoft/azure/functions/worker/broker/JavaFunctionBroker.java [hunk 20]
```diff
@@ -24,6 +26,13 @@
  * reflection, and invoke them at runtime. Thread-Safety: Multiple thread.
  */
 public class JavaFunctionBroker {
+
+	//TODO: build dedicate ImmutablePair class with meaningful fields.
+	private final Map<String, ImmutablePair<String, FunctionDefinition>> methods;
+	private final ClassLoaderProvider classLoaderProvider;
+	private String workerDirectory;
+	private volatile boolean loadMiddleware = true;
```
Comment (trask) [inline src/main/java/com/microsoft/azure/functions/worker/broker/JavaFunctionBroker.java]:
```suggestion
	private final AtomicBoolean invocationChainFactoryInitialized = new AtomicBoolean();
```

File: src/main/java/com/microsoft/azure/functions/worker/broker/JavaFunctionBroker.java [hunk 21]
```diff
@@ -33,15 +42,47 @@ public void loadMethod(FunctionMethodDescriptor descriptor, Map<String, BindingI
...
 		addSearchPathsToClassLoader(descriptor);
+		loadMiddleware();
 		FunctionDefinition functionDefinition = new FunctionDefinition(descriptor, bindings, classLoaderProvider);
 		this.methods.put(descriptor.getId(), new ImmutablePair<>(descriptor.getName(), functionDefinition));
 	}
 
+	private void loadMiddleware() {
+		if (loadMiddleware) {
+			synchronized (JavaFunctionBroker.class){
+				if (loadMiddleware) {
```
Comment (trask) [inline src/main/java/com/microsoft/azure/functions/worker/broker/JavaFunctionBroker.java]: if (!middlewareLoaded.getAndSet(true)) {

File: src/main/java/com/microsoft/azure/functions/worker/broker/JavaFunctionBroker.java [hunk 22]
```diff
@@ -33,15 +42,47 @@ public void loadMethod(FunctionMethodDescriptor descriptor, Map<String, BindingI
...
+					ArrayList<FunctionWorkerMiddleware> middlewares = new ArrayList<>();
+					try {
+						//ServiceLoader will use thread context classloader to verify loaded class
+						Thread.currentThread().setContextClassLoader(classLoaderProvider.createClassLoader());
+						ServiceLoader<FunctionWorkerMiddleware> middlewareServiceLoader = ServiceLoader.load(FunctionWorkerMiddleware.class);
+						for (FunctionWorkerMiddleware middleware : middlewareServiceLoader) {
+							middlewares.add(middleware);
+							WorkerLogManager.getSystemLogger().info("Load middleware " + middleware.getClass().getSimpleName());
```
Comment (trask) [inline src/main/java/com/microsoft/azure/functions/worker/broker/JavaFunctionBroker.java]:
```suggestion
							WorkerLogManager.getSystemLogger().debug("Load middleware " + middleware.getClass().getSimpleName());
```

File: src/main/java/com/microsoft/azure/functions/worker/broker/JavaFunctionBroker.java [hunk 23]
```diff
@@ -33,15 +42,47 @@ public void loadMethod(FunctionMethodDescriptor descriptor, Map<String, BindingI
...
+	private void loadMiddleware() {
+		if (loadMiddleware) {
+			synchronized (JavaFunctionBroker.class){
+				if (loadMiddleware) {
+					ArrayList<FunctionWorkerMiddleware> middlewares = new ArrayList<>();
+					try {
+						//ServiceLoader will use thread context classloader to verify loaded class
+						Thread.currentThread().setContextClassLoader(classLoaderProvider.createClassLoader());
+						ServiceLoader<FunctionWorkerMiddleware> middlewareServiceLoader = ServiceLoader.load(FunctionWorkerMiddleware.class);
+						for (FunctionWorkerMiddleware middleware : middlewareServiceLoader) {
```
Comment (trask) [inline src/main/java/com/microsoft/azure/functions/worker/broker/JavaFunctionBroker.java]: inline the var

File: src/main/java/com/microsoft/azure/functions/worker/broker/JavaFunctionBroker.java [hunk 24]
```diff
@@ -33,15 +42,47 @@ public void loadMethod(FunctionMethodDescriptor descriptor, Map<String, BindingI
...
+				}
+			}
+		}
+	}
+
+	private void loadFunctionExecutionMiddleWare(ArrayList<FunctionWorkerMiddleware> middlewares) {
+		FunctionExecutionMiddleware functionExecutionMiddleware = new FunctionExecutionMiddleware(
+				JavaMethodExecutors.createJavaMethodExecutor(this.classLoaderProvider.createClassLoader()));
+		middlewares.add(functionExecutionMiddleware);
+		WorkerLogManager.getSystemLogger().info("Load last middleware: FunctionExecutionMiddleware");
```
Comment (trask) [inline src/main/java/com/microsoft/azure/functions/worker/broker/JavaFunctionBroker.java:73]:
```suggestion
		WorkerLogManager.getSystemLogger().debug("Load last middleware: FunctionExecutionMiddleware");
```

File: src/main/java/com/microsoft/azure/functions/worker/broker/JavaFunctionBroker.java [hunk 25]
```diff
@@ -33,15 +42,47 @@ public void loadMethod(FunctionMethodDescriptor descriptor, Map<String, BindingI
...
+						ServiceLoader<FunctionWorkerMiddleware> middlewareServiceLoader = ServiceLoader.load(FunctionWorkerMiddleware.class);
+						for (FunctionWorkerMiddleware middleware : middlewareServiceLoader) {
+							middlewares.add(middleware);
+							WorkerLogManager.getSystemLogger().info("Load middleware " + middleware.getClass().getSimpleName());
+						}
+					} finally {
+						Thread.currentThread().setContextClassLoader(ClassLoader.getSystemClassLoader());
+					}
+					loadFunctionExecutionMiddleWare(middlewares);
+					loadMiddleware = false;
```
Comment (trask) [inline src/main/java/com/microsoft/azure/functions/worker/broker/JavaFunctionBroker.java]:
```suggestion
					this.invocationChainFactory = ...
```

File: src/main/java/com/microsoft/azure/functions/worker/broker/JavaFunctionBroker.java [hunk 26]
```diff
@@ -33,15 +42,47 @@ public void loadMethod(FunctionMethodDescriptor descriptor, Map<String, BindingI
...
+				}
+			}
+		}
+	}
+
+	private void loadFunctionExecutionMiddleWare(ArrayList<FunctionWorkerMiddleware> middlewares) {
```
Comment (trask) [inline src/main/java/com/microsoft/azure/functions/worker/broker/JavaFunctionBroker.java]:
```suggestion
	private static void addFunctionExecutionMiddleware(ArrayList<FunctionWorkerMiddleware> middlewares) {
```

**Review by trask — COMMENTED** [review]
from joint review with @kaibocai @shreyas-gopalakrishna

File: src/main/java/com/microsoft/azure/functions/worker/binding/ExecutionContextDataSource.java [hunk 27]
```diff
@@ -64,4 +92,82 @@ public MethodBindInfo getMethodBindInfo() {
...
+            if(annotation.annotationType().getSimpleName().equals(name)){
+                return true;
+            }
+        }
+        return false;
+    }
+
+    // TODO: Refactor the code in V5 to make resolve arguments logics before middleware invocation.
+    //  For now only support String type data.
+    private static Map<String, Object> buildParameterValues(List<ParameterBinding> parameterBindings){
```
Comment (trask) [inline src/main/java/com/microsoft/azure/functions/worker/binding/ExecutionContextDataSource.java]:
```suggestion
    // for now only supporting String parameter values mapped to String values
    private static Map<String, Object> resolveParameterValuesForMiddleware(List<ParameterBinding> parameterBindings){
```

File: src/main/java/com/microsoft/azure/functions/worker/binding/ExecutionContextDataSource.java [hunk 28]
```diff
@@ -18,14 +27,31 @@ public final class ExecutionContextDataSource extends DataSource<ExecutionContex
...
+            name = "req",
+            methods = {HttpMethod.GET, HttpMethod.POST},
+            authLevel = AuthorizationLevel.ANONYMOUS)
+    HttpRequestMessage<Optional<String>> request,
+    Here name will be the "req".
+
+    Value is java.lang.reflect.Parameter type
+     */
+    private final Map<String, Parameter> parameterDefinitions;
+    private final Map<String, Object> parameterValues;
```
Comment (trask) [inline src/main/java/com/microsoft/azure/functions/worker/binding/ExecutionContextDataSource.java:45]:
```suggestion
    // for now only contains String parameter values
    // and only supports grpc String -> middleware String resolution
    private final Map<String, Object> parameterValues;
```

File: src/main/java/com/microsoft/azure/functions/worker/binding/ExecutionContextDataSource.java [hunk 29]
```diff
@@ -18,14 +27,31 @@ public final class ExecutionContextDataSource extends DataSource<ExecutionContex
...
+    private final Map<String, Object> middlewareInputs = new HashMap<>();
```
Comment (trask) [inline src/main/java/com/microsoft/azure/functions/worker/binding/ExecutionContextDataSource.java]:
```suggestion
    // these are parameters provided by the middleware, which will override the host provided parameter values
    private final Map<String, Object> middlewareParameterValues = new HashMap<>();
```

File: src/main/java/com/microsoft/azure/functions/worker/binding/ExecutionContextDataSource.java [hunk 30]
```diff
@@ -64,4 +92,82 @@ public MethodBindInfo getMethodBindInfo() {
...
+        return map;
+    }
+
+    @Override
+    public Object getParameterValue(String name) {
+        return this.parameterValues.get(name);
+    }
+
+    @Override
+    public void updateParameterValue(String key, Object value) {
```
Comment (trask) [inline src/main/java/com/microsoft/azure/functions/worker/binding/ExecutionContextDataSource.java]:
```suggestion
    public void updateParameterValue(String name, Object value) {
```

File: src/main/java/com/microsoft/azure/functions/worker/binding/ExecutionContextDataSource.java [hunk 31]
```diff
@@ -64,4 +92,82 @@ public MethodBindInfo getMethodBindInfo() {
...
+    }
+
+    @Override
+    public Object getReturnValue() {
+        return this.returnValue;
+    }
+
+    @Override
+    public void setReturnValue(Object returnValue) {
+        this.returnValue = returnValue;
```
Comment (trask) [inline src/main/java/com/microsoft/azure/functions/worker/binding/ExecutionContextDataSource.java:162]:
```suggestion
        this.returnValue = returnValue;
        this.dataStore.setDataTargetValue(BindingDataStore.RETURN_NAME, returnValue);
```

File: src/main/java/com/microsoft/azure/functions/worker/binding/ExecutionContextDataSource.java [hunk 32]
```diff
@@ -64,4 +92,82 @@ public MethodBindInfo getMethodBindInfo() {
...
+    @Override
+    public void updateParameterValue(String key, Object value) {
+        this.middlewareInputs.put(key, value);
+    }
+
+    // set the return value that will be sent back to host
+    public void setHostReturnValue() {
+        this.dataStore.setDataTargetValue(BindingDataStore.RETURN_NAME, this.returnValue);
+    }
+
```
Comment (trask) [inline src/main/java/com/microsoft/azure/functions/worker/binding/ExecutionContextDataSource.java]:
```suggestion
```

File: src/main/java/com/microsoft/azure/functions/worker/broker/JavaFunctionBroker.java [hunk 33]
```diff
@@ -33,15 +44,41 @@ public void loadMethod(FunctionMethodDescriptor descriptor, Map<String, BindingI
 			throws ClassNotFoundException, NoSuchMethodException, IOException {
 		descriptor.validate();
 		addSearchPathsToClassLoader(descriptor);
+		initializeInvocationChainFactory();
 		FunctionDefinition functionDefinition = new FunctionDefinition(descriptor, bindings, classLoaderProvider);
 		this.methods.put(descriptor.getId(), new ImmutablePair<>(descriptor.getName(), functionDefinition));
 	}
 
+	private void initializeInvocationChainFactory() {
+		if (invocationChainFactoryInitialized.compareAndSet(false, true)) {
```
Comment (trask) [inline src/main/java/com/microsoft/azure/functions/worker/broker/JavaFunctionBroker.java]:
```suggestion
		if (!invocationChainFactoryInitialized.getAndSet(true)) {
```

File: src/main/java/com/microsoft/azure/functions/worker/broker/JavaFunctionBroker.java [hunk 34]
```diff
@@ -33,15 +44,41 @@ public void loadMethod(FunctionMethodDescriptor descriptor, Map<String, BindingI
...
+			try {
+				//ServiceLoader will use thread context classloader to verify loaded class
+				Thread.currentThread().setContextClassLoader(classLoaderProvider.createClassLoader());
+				for (FunctionWorkerMiddleware middleware : ServiceLoader.load(FunctionWorkerMiddleware.class)) {
+					middlewares.add(middleware);
+					WorkerLogManager.getSystemLogger().info("Load middleware " + middleware.getClass().getSimpleName());
+				}
+			} finally {
+				Thread.currentThread().setContextClassLoader(ClassLoader.getSystemClassLoader());
+			}
```
Comment (trask) [inline src/main/java/com/microsoft/azure/functions/worker/broker/JavaFunctionBroker.java]: extract to method? `loadMiddlewares`

File: src/main/java/com/microsoft/azure/functions/worker/chain/InvocationChain.java [hunk 35]
```diff
@@ -0,0 +1,23 @@
...
+
+     public InvocationChain(List<FunctionWorkerMiddleware> middlewares){
+         this.middlewareIterator = middlewares.iterator();
+     }
+
+    @Override
+    public void doNext(MiddlewareContext context) throws Exception {
+        while (middlewareIterator.hasNext()) {
+            middlewareIterator.next().invoke(context, this);
+        }
```
Comment (trask) [inline src/main/java/com/microsoft/azure/functions/worker/chain/InvocationChain.java]:
```suggestion
        if (middlewareIterator.hasNext()) {
            middlewareIterator.next().invoke(context, this);
        }

```

File: src/main/java/com/microsoft/azure/functions/worker/binding/ExecutionContextDataSource.java [hunk 36]
```diff
@@ -18,14 +27,36 @@ public final class ExecutionContextDataSource extends DataSource<ExecutionContex
...
+    Value is java.lang.reflect.Parameter type
+     */
+    private final Map<String, Parameter> parameterDefinitions;
+
+    // for now only contains String parameter values
+    // and only supports grpc String -> middleware String resolution
+    private final Map<String, Object> parameterValues;
+
+    // these are parameters provided by the middleware, which will override the host provided parameter values
+    private final Map<String, Object> middlewareParameterValues = new HashMap<>();
```
Comment (trask) [inline src/main/java/com/microsoft/azure/functions/worker/binding/ExecutionContextDataSource.java:49]:
```suggestion
    // currently the values are only Strings, but planning to support other types in the future
    private final Map<String, Object> middlewareParameterValues = new HashMap<>();
```

File: src/main/java/com/microsoft/azure/functions/worker/binding/ExecutionContextDataSource.java [hunk 37]
```diff
@@ -18,14 +27,36 @@ public final class ExecutionContextDataSource extends DataSource<ExecutionContex
...
+            authLevel = AuthorizationLevel.ANONYMOUS)
+    HttpRequestMessage<Optional<String>> request,
+    Here name will be the "req".
+
+    Value is java.lang.reflect.Parameter type
+     */
+    private final Map<String, Parameter> parameterDefinitions;
+
+    // for now only contains String parameter values
+    // and only supports grpc String -> middleware String resolution
```
Comment (trask) [inline src/main/java/com/microsoft/azure/functions/worker/binding/ExecutionContextDataSource.java]:
```suggestion
    // currently the values are only Strings (resolved from grpc String values)
    // but planning to support other types in the future
```

File: src/main/java/com/microsoft/azure/functions/worker/binding/ExecutionContextDataSource.java [hunk 38]
```diff
@@ -35,6 +66,8 @@ public ExecutionContextDataSource(String invocationId, TraceContext traceContext
         this.dataStore = dataStore;
         this.methodBindInfo = methodBindInfo;
         this.containingClass = containingClass;
+        this.parameterDefinitions = addParameters(methodBindInfo);
```
Comment (trask) [inline src/main/java/com/microsoft/azure/functions/worker/binding/ExecutionContextDataSource.java]:
```suggestion
        this.parameterDefinitions = getParameterDefinitions(methodBindInfo);
```

File: src/main/java/com/microsoft/azure/functions/worker/binding/ExecutionContextDataSource.java [hunk 39]
```diff
@@ -64,4 +97,78 @@ public MethodBindInfo getMethodBindInfo() {
...
+    private static Map<String, Parameter> addParameters(MethodBindInfo methodBindInfo){
+        Map<String, Parameter> map = new HashMap<>();
+        for (ParamBindInfo paramBindInfo : methodBindInfo.getParams()) {
+            map.put(paramBindInfo.getName(), paramBindInfo.getParameter());
+        }
+        return map;
+    }
+
+    @Override
+    public Optional<String> getParameterName(String annotationType){
```
Comment (trask) [inline src/main/java/com/microsoft/azure/functions/worker/binding/ExecutionContextDataSource.java]: to make it clearer what should be passed in
```suggestion
    public Optional<String> getParameterName(String annotationSimpleClassName){
```

Comment (trask) [reply src/main/java/com/microsoft/azure/functions/worker/binding/ExecutionContextDataSource.java]: let's discuss the usage of `Optional` in tomorrow's API review

File: src/main/java/com/microsoft/azure/functions/worker/binding/ExecutionContextDataSource.java [hunk 40]
```diff
@@ -64,4 +97,78 @@ public MethodBindInfo getMethodBindInfo() {
...
+        for (Map.Entry<String, Parameter> entry : this.parameterDefinitions.entrySet()){
+            if (isTargetTrigger(entry.getValue(), annotationType)){
+                return Optional.of(entry.getKey());
+            }
+        }
+        return Optional.empty();
+    }
+
+    private static boolean isTargetTrigger(Parameter parameter, String name){
```
Comment (trask) [inline src/main/java/com/microsoft/azure/functions/worker/binding/ExecutionContextDataSource.java]:
```suggestion
    private static boolean hasAnnotation(Parameter parameter, String annotationSimpleClassName){
```

File: src/main/java/com/microsoft/azure/functions/worker/binding/ExecutionContextDataSource.java [hunk 41]
```diff
@@ -64,4 +97,78 @@ public MethodBindInfo getMethodBindInfo() {
...
+        this.dataStore.setDataTargetValue(BindingDataStore.RETURN_NAME, this.returnValue);
+    }
+
+    public Optional<BindingData> getBindingData(String paramName, Type paramType) {
+        Optional<BindingData> argument;
+        Object inputValue = this.middlewareParameterValues.get(paramName);
+        if (inputValue != null) {
+            argument = Optional.of(new BindingData(inputValue));
+        }else{
+            argument = dataStore.getDataByName(paramName, paramType);
```
Comment (trask) [inline src/main/java/com/microsoft/azure/functions/worker/binding/ExecutionContextDataSource.java]:
```suggestion
            return Optional.of(new BindingData(inputValue));
        }else{
            return dataStore.getDataByName(paramName, paramType);
```

File: src/main/java/com/microsoft/azure/functions/worker/binding/ExecutionContextDataSource.java [hunk 42]
```diff
@@ -64,4 +98,79 @@ public MethodBindInfo getMethodBindInfo() {
...
+
+    @Override
+    public void setReturnValue(Object returnValue) {
+        this.returnValue = returnValue;
+        // set the return value that will be sent back to host
+        this.dataStore.setDataTargetValue(BindingDataStore.RETURN_NAME, this.returnValue);
+    }
+
+    public Optional<BindingData> getBindingData(String paramName, Type paramType) {
+        Optional<BindingData> argument;
```
Comment (trask) [inline src/main/java/com/microsoft/azure/functions/worker/binding/ExecutionContextDataSource.java]:
```suggestion
```

**Review by trask — APPROVED** [review]
we can discuss/review the `Optional` usage on the API pr

File: src/main/java/com/microsoft/azure/functions/worker/binding/ExecutionContextDataSource.java (hunk 42 above)
Comment (kaibocai) [reply src/main/java/com/microsoft/azure/functions/worker/binding/ExecutionContextDataSource.java]: Oops missed this one :)

**Pull request merged**